---
title: "NWS_Metagenome_PHAGE"
author: "Jacob Agerbo Rasmussen"
date: "1/31/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Setting dependencies
```{r dependencies, echo=FALSE, warning=FALSE}
library(readxl);packageVersion("readxl")
library(tidyverse);packageVersion("tidyverse")
library(ggplot2);packageVersion("ggplot2")
library(ggpubr);packageVersion("ggpubr")
library(ggfortify);packageVersion("ggfortify")
library(rstatix);packageVersion("rstatix")
library(phyloseq);packageVersion("phyloseq")
library(ape);packageVersion("ape")
library(metacoder);packageVersion("metacoder")
library(cowplot);packageVersion("cowplot")
library(boral);packageVersion("boral")
library(ggboral);packageVersion("ggboral")
library(hilldiv);packageVersion("hilldiv")
library(ADImpute);packageVersion("ADImpute")
library(vegan);packageVersion("vegan")
library(phylosmith);packageVersion("phylosmith")
library(RColorBrewer);packageVersion("RColorBrewer")
library(ComplexHeatmap);packageVersion("ComplexHeatmap")
set.seed(1) # This makes the plot appear the same each time it is run 
n <- 60
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
```

Loading MAG data
```{r load data, echo=FALSE}
# df
df <- read.csv("raw_vOTU_table.txt", sep = "\t", row.names = 1)
Summary <- read.csv("final-viral-score.tsv", sep = "\t")

# tax
tax <- read.csv("Phage_taxonomy_lineage.txt", sep = "\t", row.names = 1)
tax <- as.matrix(tax)

# select only vOTUs with taxa
df <- df[match(rownames(tax), rownames(df)),]
identical(rownames(df), rownames(tax))
### TMP normalisation
Summary <- Summary[match(rownames(df), Summary$seqname),]
identical(rownames(df), Summary$seqname)
size <- data.frame(vOTU = Summary$seqname, Length = Summary$length)
colnames(size) <- c("hgnc_symbol", "transcript_length")
df.tmp <- NormalizeTPM(df, tr_length = size, scale = 1e+06) # the scale is just a constant, which can be changed, i added a million, so the total sum for each sample i 1.
colSums(df.tmp) # check Sample sum
# md
md <- read_xlsx("metadata.xlsx")
# Tidy up md
md <- as.matrix(md)
rownames(md) <- md[,1]
md <- as.data.frame(md)

# 
colnames(df.tmp) <- md$`Sample Name`
## Make physeq object
physeq.tmp <- phyloseq(otu_table(df.tmp,taxa_are_rows=TRUE),
               tax_table(tax),
               sample_data(md))

#
random_tree = rtree(ntaxa(physeq.tmp), rooted=TRUE, tip.label=taxa_names(physeq.tmp))
physeq.tmp = merge_phyloseq(physeq.tmp, random_tree)
```


Check saturation of MAGs  per depth for each sample (rarefraction curve)
```{r Rarefraction curves,echo=FALSE}
#pdf("Rarefraction.pdf", width = 9, height = 6)

df.rare <- round(df,0) # make the data to rounded percentage
vOTU.raremax <- min(rowSums(t(df.rare)))
rarecurve(t(df.rare), step = 1000, col = "black", cex = 0.6, label = TRUE, xlab = "Summed Reads", ylab = "vOTUs")
#dev.off()
```

Lets make some visual description of the composition of MAGs (especially between wild and captive individuals)
```{r Make barplots, fig.height=4, fig.width=12}
set.seed(1) # This makes the plot appear the same each time it is run 
p_gut = plot_bar(physeq.tmp, "Sample", fill = "Family") + 
  geom_bar(aes(color=Family, fill=Family), stat="identity", position="stack") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90)) +
  scale_fill_manual(values = col_vector) +
  scale_color_manual(values = col_vector) +
  labs(x = "Individuals") +
  labs(y = "Relative vOTU Abundance")

#p_gut + facet_wrap(~Location, scales = "free", nrow = 1)
#p_gut + facet_wrap(~Sex, scales = "free", nrow = 1)
#p_gut + facet_wrap(~Stomach_Content, scales = "free", nrow = 1)


GP.ord <- ordinate(physeq.tmp, "PCoA", "unifrac", weighted=TRUE)
pcoa = plot_ordination(physeq.tmp, GP.ord, type="samples",
                     color="Wild_OR_Captive",
                     label="ID",
                     title="vOTU composition of wild avians") +
  geom_point(size=5) + theme_bw() +
  scale_fill_manual(values = col_vector) +
  scale_color_manual(values = col_vector) + ggrepel::geom_label_repel(label = md$`Sample Name`)


```


```{r Make barplots PDF, fig.height=4, fig.width=12}
# Make PDF
#pdf("vOTU_Composition.pdf", height = 6, width = 12)
plot_grid(pcoa,p_gut, nrow = 2, labels = 'AUTO')
#dev.off()
```
Lets make some richness/diversity assessment of gut microbiota

```{r Make diversity}
## Make physeq object
df.rich <- round(df.tmp*1000000,0) # make TMP normalised data to round percentage
physeq.rich <- phyloseq(otu_table(df.rich,taxa_are_rows=TRUE),
               tax_table(tax),
               sample_data(md))
#
random_tree = rtree(ntaxa(physeq.rich), rooted=TRUE, tip.label=taxa_names(physeq.rich))
physeq.rich = merge_phyloseq(physeq.rich, random_tree)


#
```
```{r Richness plot}
#pdf("vOTU_Richness.pdf", width = 12, height = 4)
plot_richness(physeq.rich, x="Wild_OR_Captive", color="Wild_OR_Captive", measures = c("Observed","Shannon", "Simpson")) + 
  theme_cowplot() + 
  geom_point(size=5) + 
  xlab("Wild_OR_Captive") +
  scale_fill_manual(values = col_vector) +
  scale_color_manual(values = col_vector) + xlab("") +
  geom_boxplot(alpha=0.6)+ 
  theme(legend.position="none", axis.text.x=element_text(angle=45,hjust=1,vjust=1,size=12))
#dev.off()
```

```{r Clean up temp}
rm(list = ls(all.names = TRUE)) #will clear all objects includes hidden objects.
gc() #free up memrory and report the memory usage.
```

```{r Co association between vOTUs}
df <- read.csv("raw_vOTU_table.txt", sep = "\t", row.names = 1)
Summary <- read.csv("final-viral-score.tsv", sep = "\t")

# tax
vOTU.tax <- read.csv("Phage_taxonomy_lineage.txt", sep = "\t", row.names = 1)

# select only vOTUs with taxa
df <- df[match(rownames(vOTU.tax), rownames(df)),]
identical(rownames(df), rownames(vOTU.tax))
### TMP normalisation
Summary <- Summary[match(rownames(df), Summary$seqname),]
identical(rownames(df), Summary$seqname)
size <- data.frame(vOTU = Summary$seqname, Length = Summary$length)
colnames(size) <- c("hgnc_symbol", "transcript_length")
vOTU.tmp <- NormalizeTPM(df, tr_length = size, scale = 1e+06) # the scale is just a constant, which can be changed, i added a million, so the total sum for each sample i 1.
colSums(vOTU.tmp) # check Sample sum
# md
md <- read_xlsx("metadata.xlsx")
# Tidy up md
md <- as.matrix(md)
rownames(md) <- md[,1]
md <- as.data.frame(md)
# 
colnames(vOTU.tmp) <- md$`Sample Name`

rm(df,size,Summary)
```
```{r Co association between vOTUs}
MAGs <- as.matrix(read_excel("MAGs/mean_coverage_Q2Q3.xlsx"))
MAG.tax <- read_excel("MAGs/Tax.xlsx")
rownames(MAGs) <- MAGs[,1]
MAGs <- MAGs[,-c(1)]
class(MAGs) <- "numeric"
bins_summary <- read_excel("MAGs/bins_summary.xlsx")

size <- data.frame(MAG = bins_summary$bins, Length = bins_summary$total_length)
colnames(size) <- c("hgnc_symbol", "transcript_length")
MAGs.tmp <- NormalizeTPM(MAGs, tr_length = size, scale = 1e+06) # the scale is just a constant, which can be changed, i added a million, so the total sum for each sample i 1.
```

```{r}
# filter low variant OTUs out
#  First find the desired quantile breaks for the entire matrix
qt <- quantile(MAGs.tmp , probs = c(0.10,0.90) )
#Taking out OTUs fall outside the qt breaks
rows <- apply( MAGs.tmp , 1 , function(x) any( x < qt[1] | x > qt[2] ) )
#And subsetting on this vector
MAGs.tmp.filtered <- MAGs.tmp[ rows , ]

# filter low variant vOTUs out
#  First find the desired quantile breaks for the entire matrix
qt <- quantile(vOTU.tmp , probs = c(0.10,0.90) )
#Taking out vOTUs fall outside the qt breaks
rows <- apply(vOTU.tmp , 1 , function(x) any( x < qt[1] | x > qt[2] ) )
vOTUs.tmp.filtered <- vOTU.tmp[ rows , ]
```
```{r Correlate filtered data}
# Correlate data
mat_data <- psych::corr.test(t(vOTUs.tmp.filtered), t(MAGs.tmp.filtered), use = "pairwise",method="pearson",adjust="bonferroni", alpha=0.05,ci=T)
p <- mat_data$p.adj
mat_data <- mat_data$r
#Summarize correlations
sum(table(mat_data[mat_data > 0.5]))
sum(table(mat_data[mat_data < -0.5]))
sum(table(p[p < 0.05]))
```


```{r Plot Correlations, fig.height=22, fig.width=10}
set.seed(1) # This makes the plot appear the same each time it is run 
n <- 60
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
## make a corrplot
# filter out non-significant correlations
rows <- apply( p , 1 , function(x) any( x < 0.05) )
table(rows)
columns <- apply( p , 2 , function(x) any( x < 0.05) )
table(columns)
mat_data.filtered <- mat_data[rows,columns]

vOTU.tax.filtered <- vOTU.tax[order(match(rownames(mat_data.filtered),rownames(vOTU.tax))),]

#Clustering of PAMs
set.seed(1)
pa_r = cluster::pam(mat_data.filtered, k = 8)
pa_c = cluster::pam(t(mat_data.filtered), k = 7)
#
MAG_info.filtered <- MAG.tax[order(match(colnames(mat_data.filtered),MAG.tax$bins)),]
MAG_tax_info <- sapply(MAG_info.filtered, function(x) n_distinct(x))

col = list(MAG_Genus = setNames(col_vector[1:MAG_tax_info["Genus"]], unique(MAG_info.filtered$Genus)),
           MAG_Phylum = setNames(col_vector[1:MAG_tax_info["Phylum"]], unique(MAG_info.filtered$Phylum)))
col$MAG_Genus[["Unknown"]] <- "grey20"
col$MAG_Family[["Unknown"]] <- "grey20"
# Annotate side graphs in heatmap

ha <- HeatmapAnnotation(
  MAG_Genus = MAG_info.filtered$Genus, 
  MAG_Phylum = MAG_info.filtered$Phylum, 
  col = col
)
# Make colour codes for HA
tax_info <- sapply(vOTU.tax.filtered, function(x) n_distinct(x))
names(tax_info) <- colnames(vOTU.tax.filtered)
ha_col <- list()
ha_col[[1]] = setNames(RColorBrewer::brewer.pal(name = "Set1", n = tax_info["Class"]), unique(vOTU.tax.filtered$Class))
ha_col[[1]][ha_col$Tax == "Unclassified"] <- "Black" 
names(ha_col) <- "Tax"
ha_col$Tax[["Unclassified"]] <- "grey20"
#
ha_mix_right = HeatmapAnnotation(Tax = vOTU.tax.filtered$Class, which = "row", width = unit(1.5, "cm"), col = ha_col)
pdf("vOTU_MAG_corr.pdf", height=14, width=10)
col_fun = circlize::colorRamp2(c(-1, 0, 1), c("#db3030", "#affabf", "#08a377"))
Heatmap(mat_data.filtered, col = col_fun,
        top_annotation = ha,
        split = paste0("pam", pa_r$clustering),
        column_split = paste0("", pa_c$clustering), 
        #row_labels = c(ifelse(hm.Annotation$CF_class != "no matches", hm.Annotation$CF_class, "")),
        #column_labels = hm.md$Individuals,
        row_dend_side = "right",
        row_names_side = c("left"),
        width = 1) + ha_mix_right
dev.off()
```