# Metagenomic analysis of western capercaillies

This Github repository includes the bioinformatic workflow and the analytical framework to generate results that are reported in Rasmunssen & Chua (2023) (unpublished) and are central to its main claims.

## Background

Birds are a taxonomically diverse group with over 10,500 documented living species (Jarvis et al. 2014; Gill and Donsker 2019). They occupy various habitat types and vary widely in their diet (Waite and Taylor 2015). Some birds are generalists, consuming a variety of food and can live in a broad range of habitats, while others are specialists with restricted diets and occupy specific habitat niches (Dos Anjos et al. 2019). Therefore, avian guts harbour a diverse array of microbes, some of which are crucial for host nutrition and immunity to cope with the vast diversity of dietary needs (Waite and Taylor 2014).

Currently, little is known about the function of wild avian gut microbiome genes or their metabolic interplay with their avian host (Cao et al., 2020). Expanding this knowledge to wild avians will be valuable in resolving patterns of influence that diet have on gut microbiota. This could be useful as a wild reference on how diet modification influences the microbial community and host fitness for industry or for conservation purposes such as reintroducing captive or bred species to repopulate wild ones.

It could be critical to use metagenomics to discover potential wild avian host hotspots of antibiotic-resistance bacteria and their associated antibiotic resistance genes (ARGs) (Bonnedahl et al. 2009; Cao et al. 2020; Cole et al. 2005; Tsubokura et al. 1995). The prevalence of antibiotic-resistance bacteria in wild avians poses a threat to both human and animal health globally (Shobrak and Abo-Amer 2014). Hence, early detection of antibiotic-resistance bacteria and knowledge of its wild avian hosts is crucial for the planning of appropriate strategies to manage this threat.

In this study, we present a comprehensive analysis of wild and captive western capercaillies (Tetrao urogallus), by combining genome-resolving metagenomics, metagenomic-derived viromes, and the mining of antibiotic-resistance genes. We investigate how lifestyle affects the different metagenomic layers and if the wild capercaillies hold potential hotspots for ARGs.

## Results


## Conclusion
